### PR TITLE
Re-enable gymnasium-atari

### PR DESCRIPTION
**Pull request by h-vetinari**
Was disabled for #9

### CODE REVIEW

**Comment by h-vetinari** [comment]
@pseudo-rnd-thoughts 
I'd like to be able to fix `gymnasium[atari]`, because we're going to need it for ray 2.3

You mentioned that this is injected somehow in the build steps upstream? Can we replicate those here?

**Comment by pseudo-rnd-thoughts** [comment]
@h-vetinari Im not sure I understand what you mean.
In Gymnasium v0.26, it does not have direct atari support due to ale-py not adding gymnasium envs. 
However, this was changed in v0.27 to use shimmy which includes the atari envs. 

At least to install atari, it is the same pip install process, `pip install gymnasium[atari]`, however, in v0.27, this installs shimmy[atari] which then installs ale-py

**Comment by h-vetinari** [comment]
> At least to install atari, it is the same pip install process, `pip install gymnasium[atari]`, ...

So what do you get when installing that, aside from the extra dependencies, and can we test that it works correctly? (e.g. an import)

For background, the main branch of Ray currently specifies `gymnasium[mujoco,atari]==0.26.3`. currently we disabled that because the old imports that gym used to have don't work anymore - this sounds intentional I guess...?

**Comment by pseudo-rnd-thoughts** [comment]
> So what do you get when installing that, aside from the extra dependencies, and can we test that it works correctly? (e.g. an import)

Yes it should be possible to run `gymnasium.make("ALE/Pong-v5")` for v0.27, for v0.26 then you need to use the gym compatibility wrapper `gymnasium.make("GymV26Environment-v0", env_id="ALE/Pong-v5")` or change the install to use `shimmy[atari]`
 
> For background, the main branch of Ray currently specifies `gymnasium[mujoco,atari]==0.26.3`. currently we disabled that because the old imports that gym used to have don't work anymore - this sounds intentional I guess...?

If I understand then I think this is related to the issue below. I wonder if you update to gymnasium==0.27.0 if this issue goes away?

**Comment by h-vetinari** [comment]
> > For background, the main branch of Ray currently specifies `gymnasium[mujoco,atari]==0.26.3`. currently we disabled that because the old imports that gym used to have don't work anymore - this sounds intentional I guess...?
> 
> If I understand then I think this is related to the issue below. I wonder if you update to gymnasium==0.27.0 if this issue goes away?

What I meant is that ray hard-pins 0.26.3, and expects whatever `gymnasium[atari]` does to work (or at least enough for their purpose). I don't know how substantial the changes are from 0.26->0.27, but we'd have to ask the ray maintainers to unpin / upgrade this (or do it on the feedstock, which is even less desirable because less thoroughly tested).

My point is, I want to be able to switch on the currently disabled output `gymnasium-atari`, _for 0.26.3_, but I don't know how to test that it's working correctly (aside from bringing in the required dependencies). Note that ray [also](https://github.com/ray-project/ray/blame/master/python/requirements/ml/requirements_rllib.txt#L6-L8) requires `gym`, so it seems they're aware of the limitations of 0.26.3.

I guess that means in any case that we'll need a gym-branch and publish gym 0.26.2, and switch on the additional `gymnasium-atari` output here but without testing the import. WDYT?

**Comment by pseudo-rnd-thoughts** [comment]
Ahh I understand what you mean now. 

That is quite annoying. Could we just test locally for v0.26.3 as this should be the only version where there are an issue then use the automated testing in v0.27?

**Comment by ChristofKaufmann** [comment]
I tried to do a local build of `gymnasium-atari` and it worked after I removed the [import test](https://github.com/conda-forge/gymnasium-feedstock/blob/b8964043f8df671c50e94e340273133e9485a68f/recipe/meta.yaml#L95). The import test appears wrong to me, since I cannot import `gymnasium.envs.atari` even with a pure pip installation.

However, to make it work, I had to install `shimmy[atari]` and `autorom[accept-rom-license]` via pip. The `autorom[accept-rom-license]` package is more or less an optional dependency, but `shimmy[atari]` is a requirement. This is also the case for the PyPI package, see [here](https://github.com/Farama-Foundation/Gymnasium/blob/e0cd42f77504060e770ab52932bf7eba45ff1976/pyproject.toml#L39). So, before re-enabling `gymnasium-atari`, we need to add `shimmy-atari` to conda-forge, right?

**Review by pseudo-rnd-thoughts — COMMENTED** [review]
Im attempting https://github.com/conda-forge/gymnasium-feedstock/pull/27 and saw line 88
```
imports:
        - gymnasium.envs.atari
```

The problem is that this should fail, atari environment no longer reside in gymnasium and ale-py was changed in v0.8 I believe to not install into gym either. 

Previously, ale-py would install then copy some files into `gym.envs.atari`, however, due to how dodgy this was, it was removed and all files exist in ale-py (and shimmy for gymnasium).

The problem is that this means the macos cases should fail, this to my knowledge, shouldn't work at all. 
I mean be wrong in my understanding of atari, but does anyone understand why the macos currently passes

**Comment by ChristofKaufmann** [comment]
@pseudo-rnd-thoughts Yes, you have in [line 85/87](https://github.com/conda-forge/gymnasium-feedstock/pull/27/files#diff-f3725a55bf339595bf865fec73bda8ac99f283b0810c205442021f29c06eea9aL85) `skip: true`, so this part including tests is completely skipped. If you or @h-vetinari wants I can contribute to #16 or #27. I thought, the focus of #27 is to add tests and not to re-enable atari, which can be done independently in #16. Just give me a pointer where and how I could help.

**Comment by ChristofKaufmann** [comment]
So, I tried a few things and the segmentation error seems not to stem from gymnasium. When I build and install the package locally, install a lot of dependencies and call `pytest -v tests/`, the tests pass (a few tests are skipped and some warnings occur).

@h-vetinari: When you change the following two points, I think this can be merged:

- Add [`shimmy-atari`](https://anaconda.org/conda-forge/shimmy-atari) as run dependency of `gymnasium-atari`.
- Remove test section of `gymnasium-atari`, since `gymnasium.envs.atari` cannot be imported.

**Comment by pseudo-rnd-thoughts** [comment]
@h-vetinari Is it possible to see if this works with the changes in #28?

**Comment by h-vetinari** [comment]
> @h-vetinari Is it possible to see if this works with the changes in #28?

I've lost track if atari should already work with 0.27 or needs 0.28? In any case, I can rebase this PR.

**Comment by pseudo-rnd-thoughts** [comment]
Atari should work with 0.27 through shimmy[atari]

**Comment by h-vetinari** [comment]
> Atari should work with 0.27 through shimmy[atari]

`shimmy-atari` is present in the environment now, but `import gymnasium.envs.atari` still fails.

**Comment by ChristofKaufmann** [comment]
> import gymnasium.envs.atari still fails.

This is expected. See [this comment](https://github.com/conda-forge/gymnasium-feedstock/pull/16#pullrequestreview-1373773022) above. Gymnasium's Atari environments are covered by the test suite. I think this import test can just be removed.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -118,7 +118,7 @@ outputs:
         - shimmy-atari
     test:
       imports:
-        - gymnasium.envs.atari
+        - echo "tested through test suite in gymnasium-all"
```
[the remark this answers is not in the document]
Comment (ChristofKaufmann) [reply recipe/meta.yaml:121]: Oh, sorry! My statement "Gymnasium's Atari environments are covered by the test suite." was wrong then. I must have mixed it up with shimmy's tests.

[the remark this answers is not in the document]
Comment (h-vetinari) [reply recipe/meta.yaml]: > It is actually shimmy where we do the testing as the atari code doesn't exist in gymnasium but this isn't an issue

The question then becomes whether `gymnasium-atari` should be an output of the shimmy feedstock rather than this one? Not that I understand how the various components are intertwined but just thinking out loud.

**Comment by h-vetinari** [comment]
Looks like `test_shimmy_gym_compatibility` fails with either (Python 3.11 builds everywhere)
```
        try:
>           env = env_creator(**_kwargs)
E           TypeError: _raise_shimmy_error() got an unexpected keyword argument 'env_id'
```
or (Python <3.11 on linux-aarch64, linux-ppc64le)
```
    def __getattr__(self, name):
        if name.startswith("_"):
            raise AttributeError(f"attempted to get missing private attribute '{name}'")
>       return getattr(self.env, name)
E       AttributeError: 'CartPoleEnv' object has no attribute 'render_mode'
```

File: recipe/meta.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (h-vetinari) [reply recipe/meta.yaml]: The fact that the inclusion of `shimmy-atari` in the test deps is causing a failure here actually does show that something about gymnasium-atari _is_ being covered here. So while we now have a test failure to fix (or skip), it does allay my concerns of where we should put this output (i.e. this feedstock is fine :)).

**Comment by ChristofKaufmann** [comment]
> Looks like `test_shimmy_gym_compatibility` fails with either (Python 3.11 builds everywhere)

Regarding the first issue, we are in the branch `elif gym is None:` and in this case it seems to be legacy gym, which is not installable on Python 3.11. In this case the exception `DependencyNotInstalled` is expected, but we get a different one. Not sure why, yet. It might be an incompatibility with shimmy 1.0.0. Before we had 0.2.1.

**Comment by h-vetinari** [comment]
> Regarding the first issue, we are in the branch `elif gym is None:` and in this case it seems to be legacy gym, which is not installable on Python 3.11. In this case the exception `DependencyNotInstalled` is expected, but we get a different one.

OK, sounds to me like we can ignore this test for now (meaning this gymnasium version). Any thoughts on the `CartPoleEnv` failures?

**Comment by h-vetinari** [comment]
Thanks for investigating!

> Just replace the contents of `recipe/patches/0002-fix-compatibility-version-mismatch.patch` with:

That patch seems to be working for all the python<3.11 versions. We could have two different patches, but again I think in this case we should just skip the test on python 3.11. The second issue does look a tad more serious AFAICT, but I could also be convinced to skip it for now and revisit for 0.28:
```
E       AttributeError: 'CartPoleEnv' object has no attribute 'render_mode'
```

**Comment by pseudo-rnd-thoughts** [comment]
For the CartPole issue, I suspect the issue is that an old gym version was installed where `render_mode` doesn't exist yet. The reason this doesn't occur for 3.11 is that the older gym versions are incompatible with it 
`    gym:                       0.22.0-py39hdae49d8_0          conda-forge` on line 2341

I saw this going through the logs - `WARNING:conda_build.build:No files or script found for output gymnasium-all` is this expected?

**Comment by h-vetinari** [comment]
> For the CartPole issue, I suspect the issue is that an old gym version was installed where render_mode doesn't exist yet. 

What do you think about skipping that test for now and re-evaluating for 0.28?

> I saw this going through the logs - `WARNING:conda_build.build:No files or script found for output gymnasium-all` is this expected?

Yes, this is expected. `gymnasium-all` is not being "built" in any way, it's just a metapackage that is only made up of its dependencies.

**Comment by ChristofKaufmann** [comment]
> That patch seems to be working for all the python<3.11 versions. We could have two different patches, but again I think in this case we should just skip the test on python 3.11.

I would agree with you, but maybe the test authors wanted occasionally to test, if the error message is still good? The [code in the test](https://github.com/Farama-Foundation/Gymnasium/blob/v0.27.1/tests/envs/test_compatibility.py#LL164-L171) is

```python
elif gym is None:
            with pytest.raises(
                DependencyNotInstalled,
                match=re.escape(
                    "No module named 'gym' (Hint: You need to install gym with `pip install gym` to use gym environments"
                ),
            ):
              gymnasium.make("GymV22Environment-v0", env_id="CartPole-v1")
```

To skip the test if gym is missing, you could just use:

```python
elif gym is None:
    pytest.skip("Cannot test without gym")
```

This would be easier, but it does not test the error message. So maybe it is intentional to test the message.

> For the CartPole issue, I suspect the issue is that an old gym version was installed where render_mode doesn't exist yet.

I agree. The issue comes from [`gymnasium.make("GymV26Environment-v0", env_id="CartPole-v1")`](https://github.com/Farama-Foundation/Gymnasium/blob/8fc9844a85ddc7d814e8494fced96b05eb10212b/tests/envs/test_compatibility.py#L181). This test does not really make sense on linux-aarch64 and linux-ppc64le, since the latest version of [gym on conda-forge](https://anaconda.org/conda-forge/gym) for these architectures is 0.22.0. However, in #28 the [test was passing](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=703300&view=logs&j=3a25022a-f08c-5c04-be7d-d279a7fc48ef&t=03fcd6e6-009d-576b-db27-27f5032302f9&l=2400) with the same version of gym.

The error in this PR is raised during creation of the environment, when shimmy tries to [access the `render_mode` attribute from the stripped environment](https://github.com/Farama-Foundation/Shimmy/blob/f1dfc55f02722e0ddbf99c6c3a7d028240dc3f8e/shimmy/openai_gym_compatibility.py#L90). The attribute does not exist in gym 0.22.0 environments. I wonder why this error was not raised in the tests of #28. There is no difference between shimmy 0.2.1 and 1.0.0 regarding the V26 compatibility layer. I also tried locally to test with gym==0.22.0 and shimmy==1.0.0 and also with gym=0.22.0 and shimmy==0.2.1 → I keep getting the error, which is fine. I don't see, what else could have been different in #28.

> What do you think about skipping that test for now and re-evaluating for 0.28?

I think this test will always fail, since gym will stay on 0.22.0 for these architectures. We need to skip/remove this line `gymnasium.make("GymV26Environment-v0", env_id="CartPole-v1")` of the tests for these architectures.

**Comment by h-vetinari** [comment]
> I think this test will always fail, since gym will stay on 0.22.0 for these architectures.

Well, I'm hoping we're not too far away from not depending on `gym` anymore _at all_, or am I misunderstanding something? If `gym` is going to be necessary longterm, we need to get it up to speed for Python 3.11 and the other arches, but I'd really like to avoid that.

> We need to skip/remove this line `gymnasium.make("GymV26Environment-v0", env_id="CartPole-v1")` of the tests for these architectures.

This is the wrong approach IMO, let's try to keep patches to tests to an absolute minimum (though we can skip individual tests through the pytest invocation if that's required for corner cases - again, a single test that's failing for reasons that are hard to work around is almost completely negligible, as opposed to the segmentation faults that we used to have, and which I'm glad we got rid of).

**Comment by pseudo-rnd-thoughts** [comment]
Checking Gymnasium, we don't actually install gym in our CI anymore, the testing is done in Shimmy. 
Therefore, if we can avoid gym being installed then the test shouldn't cause an issue I believe

**Comment by ChristofKaufmann** [comment]
> Well, I'm hoping we're not too far away from not depending on gym anymore at all, or am I misunderstanding something?

It is just an optional dependency.

> Checking Gymnasium, we don't actually install gym in our CI anymore, the testing is done in Shimmy.

Its not installed in the PyPI tests indeed. But that also means `GymV26Environment-v0` and `GymV21Environment-v0` are not covered by the tests.

> This is the wrong approach IMO, let's try to keep patches to tests to an absolute minimum

I thought of something like checking the gym version and only testing `GymV26Environment-v0` if the version is high enough. Maybe this can be done upstream (would still require a patch for now)?

**Comment by h-vetinari** [comment]
> I thought of something like checking the gym version and only testing `GymV26Environment-v0` if the version is high enough. Maybe this can be done upstream (would still require a patch for now)?

I appreciate the enthusiasm for doing things right! I'll let Mark be the judge whether a given PR to upstream gymnasium is reasonable, but as I keep saying, I strongly believe that this is not worth the effort. Note also we're not building the most recent version here yet, meaning that any changes upstream will still need backporting either upstream or here.

IMO it's just not a good investment of everyone's time to chase a single, explainable, not-indicating-a-fundamental-problem test failure. That said, if you come up with a PR that has passing CI without skipping tests, I'll take a look at it (but I'd rather invest such energy - including yours, if available - into 0.28 etc.).

**Comment by ChristofKaufmann** [comment]
@pseudo-rnd-thoughts You are right, that it is not really needed, but if you want an upstream fix, I already have working code and can make a PR.

It would replace patch 0002 in the feedstock, but I am OK, if you rather want to skip it.

**Comment by pseudo-rnd-thoughts** [comment]
OSX is failing on `CartPoleJax-v0`

**Comment by h-vetinari** [comment]
> OSX is failing on `CartPoleJax-v0`

That same job passed before, I'm not sure if the agent crashed for some reason or whether there's something flaky in the test suite. BTW, AFAICT, it failed in `tests/envs/test_rendering.py::test_render_modes[spec8]`.

**Comment by h-vetinari** [comment]
> Looked like a weird CI bug, @h-vetinari are you happy to merge and see if we can make a PR to v0.28?

Alright, let's go! :)

**Comment by h-vetinari** [comment]
> I mean with `noarch` we would not get these errors and might overlook something, but it would save a lot of CI time, if this is a concern.

For now I definitely prefer more CI coverage
